src/cluster_migrate.c

```diff
@@ -0,0 +1,2003 @@
+            }
+            return;
+        case SLOT_EXPORT_FAILOVER_GRANTED:
+            if (isSlotExportPauseTimedOut(job)) {
+                /* Note that we think this won't happen very commonly. The main source of
+                 * latency that may trigger an unpause will occur due to the time it takes
+                 * to process the incremental changes. The final REQUEST-FAILOVER handshake
+                 * will validate that the source node is still paused after this initial
+                 * handshake, and renew the pause for an additional amount of time. From this
+                 * point, we expect the takeover of the slot and gossip to be relatively quick
+                 * in steady state.
+                 *
+                 * Regardless, we log a warning and proceed with cleaning up the slot migration
+                 * job. */
+                serverLog(LL_WARNING, "Write loss risk! During slot migration %s, new owner did not "
+                                      "broadcast ownership before we unpaused ourselves. Any "
+                                      "writes we have recorded since unpausing will now be lost!",
+                          job->description);
+
+                finishSlotMigrationJob(job, SLOT_MIGRATION_JOB_FAILED,
+                                       "Unpaused before migration completed (migration may have "
+                                       "succeeded with lost writes)");
+                return;
+            }
+            /* updateSlotExportIfOwnershipChanged will close the slot migration job if it is
+             * done. */
+            updateSlotExportIfOwnershipChanged(job);
+            return;
+
+        /* Terminal states */
+        case SLOT_MIGRATION_JOB_SUCCESS:
+        case SLOT_MIGRATION_JOB_CANCELLED:
+        case SLOT_MIGRATION_JOB_FAILED:
+            return;
+        }
+    }
+}
+
+/* Reset the client and connection information associated with the job, leaving
+ * the migration related metadata. */
+void resetSlotMigrationJob(slotMigrationJob *job) {
+    /* Only one of client or conn should be set. */
+    serverAssert(!job->client || !job->conn);
+    if (job->client) {
+        freeClientAsync(job->client);
+        job->client = NULL;
+    } else if (job->conn) {
+        connClose(job->conn);
+        job->conn = NULL;
+    }
+
+    if (job->response_buf) {
+        sdsfree(job->response_buf);
+        job->response_buf = NULL;
+    }
+    if (job->description) {
+        /* Description is not needed once migration is finished */
+        sdsfree(job->description);
+        job->description = NULL;
+    }
+}
+
+void freeSlotMigrationJob(void *o) {
+    slotMigrationJob *job = o;
+    resetSlotMigrationJob(job);
+    listRelease(job->slot_ranges);
+    sdsfree(job->slot_ranges_str);
+    if (job->status_msg) {
+        sdsfree(job->status_msg);
+    }
+    if (job->response_buf) {
+        sdsfree(job->response_buf);
+    }
```

Comment:
This does not really matter, just want to save some lines
```suggestion
    if (job->status_msg) sdsfree(job->status_msg);
    if (job->response_buf) sdsfree(job->response_buf);
```

Reply:
I like it more compact. Done